analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/AvoidUnderPosting.cs

```diff
@@ -0,0 +1,143 @@
+            && method.IsControllerMethod())
+        {
+            var actionParameters = method.Parameters
+                .Select(x => x.Type)
+                .OfType<INamedTypeSymbol>()
+                .Where(x => !IgnoreType(x))
+                .SelectMany(x => RelatedTypesToExamine(x))
+                .Distinct();
+            foreach (var parameter in actionParameters)
+            {
+                CheckInvalidProperties(parameter, context, examinedTypes);
+            }
+        }
+    }
+
+    private static void CheckInvalidProperties(INamedTypeSymbol parameterType, SonarSyntaxNodeReportingContext context, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes)
+    {
+        var invalidProperties = GetAllDeclaredProperties(parameterType, examinedTypes)
+                .Where(x => !CanBeNull(x.Type)
+                    && !x.HasAttribute(KnownType.System_ComponentModel_DataAnnotations_RequiredAttribute)
+                    && !x.HasAttribute(KnownType.Microsoft_AspNetCore_Mvc_ModelBinding_Validation_ValidateNeverAttribute));
+        foreach (var property in invalidProperties)
+        {
+            var propertySyntax = property.DeclaringSyntaxReferences[0].GetSyntax();
+            context.ReportIssue(Rule, propertySyntax.GetLocation());
+        }
+    }
+
+    private static bool IgnoreType(ITypeSymbol type) =>
+        type.IsAny(IgnoredTypes)
+        || (!type.DeclaringSyntaxReferences.Any()
+            && !type.Is(KnownType.System_Collections_Generic_IEnumerable_T)
+            && !type.Implements(KnownType.System_Collections_Generic_IEnumerable_T));
+
+    private static bool CanBeNull(ITypeSymbol type) =>
+        type is ITypeParameterSymbol { HasValueTypeConstraint: false }
+        || (type.IsReferenceType && type.NullableAnnotation() != NullableAnnotation.NotAnnotated)
+        || type.IsNullableValueType();
+
+    private static IEnumerable<IPropertySymbol> GetAllDeclaredProperties(ITypeSymbol type, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes)
+    {
+        var allProperties = Enumerable.Empty<IPropertySymbol>();
+        if (type is INamedTypeSymbol namedType && examinedTypes.TryAdd(namedType, true))
+        {
+            var properties = namedType.GetMembers()
+                .OfType<IPropertySymbol>()
+                .Where(x => x.GetEffectiveAccessibility() == Accessibility.Public
+                    && x.SetMethod?.DeclaredAccessibility is Accessibility.Public);
+            foreach (var property in properties)
+            {
+                allProperties = allProperties.Append(property);
+                if (property.Type.DeclaringSyntaxReferences.Length > 0)
+                {
+                    allProperties = allProperties.Concat(GetAllDeclaredProperties(property.Type, examinedTypes));
+                }
+            }
+            ITypeSymbol[] types = [namedType.BaseType, .. namedType.TypeArguments];
```

Comment:
Can this cause an endless recursion with self-referencing type constraints?
```cs
public class C {
    public void M(MyModel model)
    {
    }
}
              
public class Model<T> where T: Model<T>
{
    public Model<T> SubModel { get; set; }
} 

public class MyModel: Model<MyModel>
{
}
```

Reply:
Does this work with type constraints (Is this supported by model binding?):
```cs
public class C {
    public void M(Model<Developer> model)
    {
    }
}
              
public class Model<T> where T: Person
{
    public T Person { get; set; }
} 

public abstract class Person
{
    public int Age { get; set; }
}

public class Developer: Person
{
    public string ProgramingLanguage {get; set; }
}
```

Reply:
I checked it and added the UTs. Model binding supports generic types with type constraints. And lucky for us, when querying `public T Person { get; set; }` in `Model<T>` the semantic model returns the properties of the `Person` class.